src/evict/evict_page.c

```diff
@@ -663,34 +663,80 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
     WT_ASSERT(session, !use_snapshot_for_app_thread || !is_eviction_thread);
 
-    if (!conn->txn_global.checkpoint_running && !WT_IS_HS(btree->dhandle) &&
-      (use_snapshot_for_app_thread || is_eviction_thread)) {
+    /* History store data is always evictable. */
+    if (WT_IS_HS(btree->dhandle)) {
+        /*
+         * FIX-ME-WT-5316: remove this when we have removed history store score and
+         * __rec_update_stable. No need to set visibility flag for history store.
+         */
+        if (!WT_SESSION_BTREE_SYNC(session))
+            LF_SET(WT_REC_VISIBLE_ALL);
+    } else {
         if (is_eviction_thread) {
-            /*
-             * Eviction threads do not need to pin anything in the cache. We have a exclusive lock
-             * for the page being evicted so we are sure that the page will always be there while it
-             * is being processed. Therefore, we use snapshot API that doesn't publish shared IDs to
-             * the outside world.
-             */
-            __wt_txn_bump_snapshot(session);
-            snapshot_acquired = true;
-
-            /*
-             * Make sure once more that there is no checkpoint running. A new checkpoint might have
-             * started between previous check and acquiring snapshot. If there is a checkpoint
-             * running, release the snapshot and fallback to global visibility checks.
-             */
-            if (conn->txn_global.checkpoint_running) {
-                __wt_txn_release_snapshot(session);
-                snapshot_acquired = false;
+            /* Eviction thread doing eviction. */
+            if (!conn->txn_global.checkpoint_running) {
+                /*
+                 * Eviction threads do not need to pin anything in the cache. We have an exclusive
+                 * lock for the page being evicted so we are sure that the page will always be there
+                 * while it is being processed. Therefore, we use snapshot API that doesn't publish
+                 * shared IDs to the outside world.
+                 */
+                __wt_txn_bump_snapshot(session);
+                snapshot_acquired = true;
+
+                /*
+                 * Make sure once more that there is no checkpoint running. A new checkpoint might
+                 * have started between previous check and acquiring snapshot. If there is a
+                 * checkpoint running, release the snapshot and fallback to global visibility
+                 * checks.
+                 *
+                 * There should be a read barrier here otherwise the second read may be optimized
+                 * away by the compiler. However, we should have already met a barrier when we bump
+                 * the snapshot so the barrier is omitted here.
+                 */
+                if (conn->txn_global.checkpoint_running) {
+                    __wt_txn_release_snapshot(session);
+                    snapshot_acquired = false;
+                    /*
```

Comment:
I appreciate that you added this comment but can we not add the same comment in 4 places? Rather at the top of the the main if statement chain add a comment.